miio/gateway/gateway.py

```diff
@@ -0,0 +1,420 @@
+    * toggle_device
+    * toggle_plug
+    * remove_all_bind
+    * list_bind [0]
+    * bind_page
+    * bind
+    * remove_bind
+
+    * self.get_prop("used_for_public") # Return the 'used_for_public' status, return value: [0] or [1], probably this has to do with developer mode.
+    * self.set_prop("used_for_public", state) # Set the 'used_for_public' state, value: 0 or 1, probably this has to do with developer mode.
+
+    * welcome
+    * set_curtain_level
+
+    * get_corridor_on_time
+    * set_corridor_light ["off"]
+    * get_corridor_light -> "on"
+
+    * set_default_sound
+    * set_doorbell_push, get_doorbell_push ["off"]
+    * set_doorbell_volume [100], get_doorbell_volume
+    * set_gateway_volume, get_gateway_volume
+    * set_clock_volume
+    * set_clock
+    * get_sys_data
+    * update_neighbor_token [{"did":x, "token":x, "ip":x}]
+
+    ## property getters
+    * ctrl_device_prop
+    * get_device_prop_exp [[sid, list, of, properties]]
+
+    ## scene
+    * get_lumi_bind ["scene", <page number>] for rooms/devices"""
+
+    def __init__(
+        self,
+        ip: str = None,
+        token: str = None,
+        start_id: int = 0,
+        debug: int = 0,
+        lazy_discover: bool = True,
+    ) -> None:
+        super().__init__(ip, token, start_id, debug, lazy_discover)
+
+        from .gateway_device import (
+            GatewayAlarm,
+            GatewayRadio,
+            GatewayZigbee,
+            GatewayLight,
+        )
+
+        self._alarm = GatewayAlarm(parent=self)
+        self._radio = GatewayRadio(parent=self)
+        self._zigbee = GatewayZigbee(parent=self)
+        self._light = GatewayLight(parent=self)
+        self._devices = {}
+        self._info = None
+
+    @property
+    def alarm(self) -> "GatewayAlarm":  # noqa: F821
+        """Return alarm control interface."""
+        # example: gateway.alarm.on()
+        return self._alarm
+
+    @property
+    def radio(self) -> "GatewayRadio":  # noqa: F821
+        """Return radio control interface."""
+        return self._radio
+
+    @property
+    def zigbee(self) -> "GatewayZigbee":  # noqa: F821
+        """Return zigbee control interface."""
+        return self._zigbee
+
+    @property
+    def light(self) -> "GatewayLight":  # noqa: F821
+        """Return light control interface."""
+        return self._light
+
+    @property
+    def devices(self):
+        """Return a dict of the already discovered devices."""
+        return self._devices
+
+    @property
+    def model(self):
+        """Return the zigbee model of the gateway."""
+        # Check if catch already has the gateway info, otherwise get it from the device
+        if self._info is None:
+            self._info = self.info()
+        return self._info.model
+
+    @command()
+    def discover_devices(self):
+        """
+        Discovers SubDevices
+        and returns a list of the discovered devices.
+        """
+
+        from .sub_device import (
```

Comment:
See my above comment wrt. location of these imports & the mapping. Maybe the subdevice class should have a static method returning an instance based on the mapping to simplify the code?

So the code below would be something like:
```
for x in range(0, len(devices_raw), 5):
    try:
        dev = SubDevice.device_from_raw(devices_raw[x : x + 5])
    except UnknownDeviceException:
        _LOGGER.warning("Unknown/unsupported device: %s", dev)
```

Reply:
I have moved large portions of the mapping from the gateway.py to the subdevice.py file.
However the exact function device_from_raw as you sugested is very specific to the info returned by the `self.get_prop("device_list")` command, therefore I would like to keep those together as it is now.

Other ways to get the subdevice list such as the Cloud or the `get_device_list` command will give back diffrently formatted info that needs to be mapped slichtly diffrent...